src/fiber/context/x86_64-microsoft.cr

```diff
@@ -0,0 +1,78 @@
+{% skip_file unless flag?(:x86_64) && flag?(:win32) %}
+
+class Fiber
+  # :nodoc:
+  def makecontext(stack_ptr, fiber_main) : Nil
+    # A great explanation on stack contexts for win32:
+    # https://cfsamson.gitbook.io/green-threads-explained-in-200-lines-of-rust/supporting-windows
+
+    # 8 registers + 2 qwords for NT_TIB + 1 parameter + 10 128bit XMM registers
+    @context.stack_top = (stack_ptr - (11 + 10*2)).as(Void*)
+    @context.resumable = 1
+
+    stack_ptr[0] = fiber_main.pointer # %rbx: Initial `resume` will `ret` to this address
+    stack_ptr[-1] = self.as(Void*)    # %rcx: puts `self` as first argument for `fiber_main`
+
+    # The following two values are stored in the Thread Information Block (NT_TIB)
+    # and are used by Windows to track the current stack limits
+    stack_ptr[-2] = @stack_bottom # %gs:0x08: Stack Bottom
+    stack_ptr[-3] = @stack        # %gs:0x10: Stack Limit
+  end
+
+  # :nodoc:
+  @[NoInline]
+  @[Naked]
+  def self.swapcontext(current_context, new_context) : Nil
+    asm("
+          pushq %rcx
+          pushq %gs:0x10    // Thread Information Block: Stack Limit
+          pushq %gs:0x08    // Thread Information Block: Stack Base
+          pushq %rdi        // push 1st argument (because of initial resume)
+          pushq %rbx        // push callee-saved registers on the stack
+          pushq %rbp
+          pushq %rsi
+          pushq %r12
+          pushq %r13
+          pushq %r14
+          pushq %r15
+          subq $$160, %rsp  // push XMM registers
+          movups %xmm6, 0x00(%rsp)
+          movups %xmm7, 0x10(%rsp)
+          movups %xmm8, 0x20(%rsp)
+          movups %xmm9, 0x30(%rsp)
+          movups %xmm10, 0x40(%rsp)
+          movups %xmm11, 0x50(%rsp)
+          movups %xmm12, 0x60(%rsp)
+          movups %xmm13, 0x70(%rsp)
+          movups %xmm14, 0x80(%rsp)
+          movups %xmm15, 0x90(%rsp)
+          movq %rsp, 0($0)  // current_context.stack_top = %rsp
+          movl $$1, 8($0)   // current_context.resumable = 1
+          movl $$0, 8($1)   // new_context.resumable = 0
+          movq 0($1), %rsp  // %rsp = new_context.stack_top
+          movups 0x00(%rsp), %xmm6 // pop XMM registers
+          movups 0x10(%rsp), %xmm7
+          movups 0x20(%rsp), %xmm8
+          movups 0x30(%rsp), %xmm9
+          movups 0x40(%rsp), %xmm10
+          movups 0x50(%rsp), %xmm11
+          movups 0x60(%rsp), %xmm12
+          movups 0x70(%rsp), %xmm13
+          movups 0x80(%rsp), %xmm14
+          movups 0x90(%rsp), %xmm15
+          addq $$160, %rsp
+          popq %r15         // pop callee-saved registers from the stack
+          popq %r14
+          popq %r13
+          popq %r12
+          popq %rsi
+          popq %rbp
+          popq %rbx
+          popq %rdi         // pop 1st argument (for initial resume)
+          popq %gs:0x08
+          popq %gs:0x10
```

Comment:
The order it popped is different from the order it stored in `makecontext`. Is it correct?

Reply:
Sorry, I didn't mean to resolve the discussion....

I believe it's a documentation error of setting rbx in makecontext (or we push rbx when we don't need to?).

We only pop up till rcx and the return address is left on the stack, I think.

Reply:
@ysbaddaden  Well, I meant these `popq %gs:0x08` and `popq %gs:0x10` two instructions , not about `rbx` and `rcx`. Sorry for including too large region. 

In `makecontext`, they store in the following order.
```
    # The following two values are stored in the Thread Information Block (NT_TIB)
    # and are used by Windows to track the current stack limits
    stack_ptr[-2] = @stack_bottom # %gs:0x08: Stack Bottom
    stack_ptr[-3] = @stack        # %gs:0x10: Stack Limit
```

Shouldn't the pop order be this?
```
  popq %gs:0x10
  popq %gs:0x08
```

Reply:
I think you're right. `GS:[0x08]` is the "high address", and `GS:[0x10]` is the "low address" and if I'm not mistaken `@stack_bottom` is the high address, and `@stack` is the low address so they're put on the initial stack in the wrong order (or we need to change both push and pop orders in `asm`).

I would just change the initial setup to:
```
stack_ptr[-2] = @stack              # %gs:0x10: Stack Limit
stack_ptr[-3] = @stack_bottom # %gs:0x08: Stack Bottom
```
@ysbaddaden might be right that we don't need to push/pop `%rcx` on every switch. I don't have the setup here to confirm that, but it might be a few cycles wasted there.

Reply:
Ah yes. They're inverted in makecontext. I totally missed that.